agent/agent_test.go

```diff
@@ -707,22 +744,67 @@ func (s *AgentSuite) TestEndTaskResponse() {
 }
 
 func (s *AgentSuite) TestOOMTracker() {
-	s.setupRunTask(defaultProjYml)
-	s.tc.project.OomTracker = true
+	projYml := `
+oom_tracker: true
+buildvariants:
+- name: mock_build_variant
+tasks: 
+ - name: this_is_a_task_name
+   commands: 
+    - command: shell.exec
+      params:
+        script: "echo hi"
+post:
+  - command: shell.exec
+    params:
+      script: "echo hi"
+`
+	s.setupRunTask(projYml)
+	s.a.opts.CloudProvider = "provider"
 	pids := []int{1, 2, 3}
 	lines := []string{"line 1", "line 2", "line 3"}
 	s.tc.oomTracker = &mock.OOMTracker{
 		Lines: lines,
 		PIDs:  pids,
 	}
 
-	_, err := s.a.runTask(s.ctx, s.tc)
+	nextTask := &apimodels.NextTaskResponse{
+		TaskId:     s.tc.task.ID,
+		TaskSecret: s.tc.task.Secret,
+		TaskGroup:  s.tc.taskGroup,
+	}
+	shouldSetupGroup := !s.tc.ranSetupGroup
+	taskDirectory := s.tc.taskDirectory
+	_, _, err := s.a.runTask(s.ctx, s.tc, nextTask, shouldSetupGroup, taskDirectory)
 	s.NoError(err)
 	s.Equal(evergreen.TaskSucceeded, s.mockCommunicator.EndTaskResult.Detail.Status)
+	// oom tracker is nil
 	s.True(s.mockCommunicator.EndTaskResult.Detail.OOMTracker.Detected)
 	s.Equal(pids, s.mockCommunicator.EndTaskResult.Detail.OOMTracker.Pids)
 }
 
+func (s *AgentSuite) TestSetupTask() {
+	var err error
+	nextTask := &apimodels.NextTaskResponse{}
+	s.setupRunTask(defaultProjYml)
+	s.tc.taskDirectory = "task_directory"
+	shouldSetupGroup, taskDirectory := s.a.finishPrevTask(s.ctx, nextTask, s.tc)
+	_, setupFailed, shouldExit, err := s.a.setupTask(s.ctx, s.ctx, s.tc, nextTask, shouldSetupGroup, taskDirectory)
+	s.False(shouldExit)
+	s.False(setupFailed)
+	s.NoError(err)
+	s.NoError(s.tc.logger.Close())
+	checkMockLogs(s.T(), s.mockCommunicator, s.tc.taskConfig.Task.Id, []string{
+		"Current command set to initial task setup (system).",
+		"Making new folder",
+		"Task logger initialized",
+		"Execution logger initialized.",
+		"System logger initialized.",
+		"Starting task 'task_id', execution 0.",
+	}, []string{panicLog})
+
+}
```

Comment:
Do we have to test the proper handling of the error conditions in `setupTask` (such as getting bad data about a task, or the context erroring while it's setting up)? Might need to tweak the mock communicator to return fake errors.

Reply:
I don't know how crucial that is because it's not doing too complex error handling and it's mostly moved from other functions that didn't test it either. The mock communicator is not set up for this to be tested easily. If you feel strongly about it I can try to add it.

Reply:
Eh, we'd ideally test it since we want to know if end in tasks system failure as expected to indicate that there's a bug (rather than the agent doing something undesirable like run the task forever, silently succeed, panic the whole agent, or something else).

If it's way too difficult though, it's alright to skip.

Reply:
yeah I think it would be difficult. We can maybe create a ticket to do it separately.

Reply:
I'll leave it to you to make the ticket.

Reply:
https://jira.mongodb.org/browse/EVG-20708